### ISSUE TITLE
preBootstrapCommands 

### ISSUE THREAD

**Issue by arielvinas** [weaveworks/eksctl#784]
**What happened?**
I'm trying to allocate resourses for the kubelet and system daemons, and also add a custom eviction force to the node, but preBootstrapCommands aren't taking effect on the node.

**What you expected to happen?**
This is my preBootstrapCommands in the yaml:
```
preBootstrapCommands:
      - "#!/bin/bash -xe"
      - "set -o xtrace"
      - "yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm"
      - "systemctl enable amazon-ssm-agent"
      - "systemctl start amazon-ssm-agent"
      - "set -a"
      - "source /etc/environment"
      - "/etc/eks/bootstrap.sh clustername \ --kubelet-extra-args '--kube-reserved=cpu=200m,memory=200Mi,ephemeral-storage=5Gi\
        --system-reserved=cpu=300m,memory=300Mi,ephemeral-storage=5Gi \
        --eviction-hard=memory.available<100Mi,nodefs.available<5%'"
```

**How to reproduce it?**
Create a nodegroup with that preBootstrapCommands and check kubelet, system and evition config and won't be the one defined in the yaml

**Anything else we need to know?**
I'm running on macOS mojave 10.14.4 with eksctl downloaded from brew as explained in the doc

**Versions**
Please paste in the output of these commands:
```
$ eksctl version
[ℹ]  version.Info{BuiltAt:"", GitCommit:"", GitTag:"0.1.31"}
$ uname -a
Darwin Macbook-air-MoniOnline 18.5.0 Darwin Kernel Version 18.5.0: Mon Mar 11 20:40:32 PDT 2019; root:xnu-4903.251.3~3/RELEASE_X86_64 x86_64
$ kubectl version 
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-19T22:12:47Z", GoVersion:"go1.12.4", Compiler:"gc", Platform:"darwin/amd64"}
```
Also include your version of `heptio-authenticator-aws`
{"Version":"0.4.0-alpha.1","Commit":"82544ec86140b7027a07890b09c2f03fca97228a"}
**Logs**
Include the output of the command line when running eksctl. If possible, eksctl should be run with debug logs. For example:
`eksctl get clusters -v 4`
Make sure you redact any sensitive information before posting.
If the output is long, please consider a Gist.

**Comment by errordeveloper** [comment]
Hi @arielvinas! Thanks for reporting this.

The idea of `preBootstrapCommands` is that it would run before the bootstrap script. Each item is ran via shell.

So here is all you actually want to do:
```
preBootstrapCommands:
     - "yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm"
      - "systemctl enable amazon-ssm-agent"
      - "systemctl start amazon-ssm-agent"
```

The following lines do nothing:
```
      - "#!/bin/bash -xe"
      - "set -o xtrace"
      - "set -a"
      - "source /etc/environment"
 ```
These would run as `sh -c  "#!/bin/bash -xe"` and `sh -c "set -o xtrace"` ...etc. So effectively a no-op.

In `eksctl` we don't use `/etc/eks/bootstrap.sh`, we have a much simpler script instead, but it doesn't support `--kubelet-extra-args`.

To set kubelet arguments you have a few options:

1) copy [`/etc/systemd/system/kubelet.service.d/10-eksclt.al2.conf`](https://github.com/weaveworks/eksctl/blob/ae4cdff698b1e7dbe2cfd7db12ede1d011508f7b/pkg/nodebootstrap/assets/10-eksclt.al2.conf) systemd drop-in, modify it and write it as e.g. `/etc/systemd/system/kubelet.service.d/20-custom.conf`
2) check if flags you want to set are supported in kubelet config file, in which case you want to modify `/etc/eksctl/kubelet.yaml`


 If you must run exactly that kind of script, you would want to pass it as a string (not array) to `overrideBootstrapCommand`.

**Comment by errordeveloper** [comment]
I think we need to document this properly, so not closing as such, but adding docs label.

**Comment by zironycho** [comment]
I have similar issue. I have to set featureGates to config [gpu pod monitoring agent](https://github.com/NVIDIA/gpu-monitoring-tools/tree/master/exporters/prometheus-dcgm/k8s/pod-gpu-metrics-exporter#prerequisites). But current eksctl couldn't set the this field.

```
[✖]  cannot override "featureGates" in kubelet config, as it's critical to eksctl functionality
```

How can I overwrite `/etc/eksctl/kubelet.yaml` this file?

**Comment by zironycho** [comment]
@errordeveloper How can I modified `/etc/eksctl/kubelet.yaml` and apply it to the ami?

**Comment by martina-if** [comment]
Hi @zironycho I made [a PR](https://github.com/weaveworks/eksctl/pull/1057) to allow overriding this field in the kubelet. It will hopefully be included in the next release. If you need it urgently, I suggest you clone the repo and compile eksctl from that branch.

**Comment by zironycho** [comment]
@martina-if wow! Thank you for your work! I will check it asap.

**Comment by shashidharrao** [comment]
@errordeveloper @martina-if  is there way i can update hostname-override using  preBootstrapCommands or adjusting 10-eksclt.al2.conf.  

hostname=$(curl -s http://169.254.169.254/latest/meta-data/hostname)
/etc/eks/bootstrap.sh ${ClusterName} --kubelet-extra-args '--hostname-override='$hostname''

here is the usecase : cluster is ready and changed the endpoint to private and using preboot scripts i am adding the nodes to domain controller and changing the hostnames to company slandered.  Now the cluster is working fine for couple of days, after the reboot it is complaining about  the hostname and it never returns the registered hostname . 

k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:47: Failed to list *v1.Pod: Unauthorized
node "ip-10-19-105-113.ec2.internal" not found
31373 kubelet.go:2251] node "ip-10-19-105-113.ec2.internal" not found
31373 kubelet.go:2251] node "ip-10-19-105-113.ec2.internal" not found
31373 kubelet.go:2251] node "ip-10-19-105-113.ec2.internal" not found
31373 kubelet.go:2251] node "ip-10-19-105-113.ec2.internal" not found

**Comment by martina-if** [comment]
Hi @shashidharrao , Have you tried using the extra kubelet configuration field in eksctl when you create the cluster? https://eksctl.io/usage/customizing-the-kubelet

If you want to change this after cluster creation I don't know any good way to do it through `eksctl` unfortunately.

**Comment by martina-if** [comment]
@arielvinas I think this was addressed as part of #1057. Please feel free to reopen this issue if you disagree or still need help with this.
